File: src/data/regionalTransportContext.js
/**
 * Regional Transport Context for the Philippines
 * Provides intelligence about regional transport corridors and connectivity
 */

/**
 * Regional transport corridor definitions
 * Groups cities by transport connectivity and infrastructure quality
 */
export const REGIONAL_TRANSPORT_CONTEXT = {
  METRO_MANILA: {
    name: "Metro Manila & Nearby",
    cities: ["Manila", "Quezon City", "Makati", "Pasig", "Taguig", "Caloocan"],
    characteristics: {
      primary_mode: "within_metro",
      infrastructure_quality: "excellent",
      tourist_friendly: true,
      notes: "Within metro - local transport only",
    },
  },

  LUZON_SOUTH: {
    name: "Southern Luzon",
    cities: ["Tagaytay", "Batangas", "Lipa", "Lucena", "Naga"],
    hub: "Manila",
    characteristics: {
      primary_mode: "bus",
      infrastructure_quality: "excellent",
      tourist_friendly: true,
      scenic_routes: true,
    },
    recommended_routes: [
      "Manila-Tagaytay",
      "Manila-Batangas",
      "Manila-Lucena",
    ],
  },

  LUZON_NORTH: {
    name: "Northern Luzon",
    cities: [
      "Baguio",
      "Banaue",
      "Sagada",
      "Vigan",
      "Laoag",
      "La Union",
      "Bontoc",
    ],
    hub: "Manila",
    characteristics: {
      primary_mode: "bus",
      infrastructure_quality: "good",
      tourist_friendly: true,
      scenic_routes: true,
      mountain_routes: true,
    },
    recommended_routes: ["Manila-Baguio", "Baguio-Sagada"],
    avoid_routes: [], // Long routes still manageable
  },

  LUZON_CENTRAL: {
    name: "Central Luzon",
    cities: ["Angeles", "Clark", "Olongapo", "Subic", "Tarlac", "Cabanatuan"],
    hub: "Manila",
    characteristics: {
      primary_mode: "bus",
      infrastructure_quality: "excellent",
      tourist_friendly: true,
    },
    recommended_routes: ["Manila-Subic", "Manila-Clark"],
  },

  MINDANAO_WEST: {
    name: "Western Mindanao Corridor",
    cities: ["Zamboanga", "Pagadian", "Dipolog", "Ozamis"],
    hub: "Zamboanga",
    characteristics: {
      primary_mode: "bus",
      infrastructure_quality: "good",
      tourist_friendly: true,
      scenic_routes: true,
      coastal_highway: true,
    },
    recommended_routes: [
      "Zamboanga-Pagadian",
      "Pagadian-Dipolog",
      "Dipolog-Ozamis",
    ],
    avoid_routes: [
      "Zamboanga-Cagayan de Oro", // Too long (11+ hours)
    ],
    notes:
      "Well-connected coastal corridor. Avoid cross-island routes to Northern Mindanao.",
  },

  MINDANAO_SOUTH: {
    name: "Southern Mindanao Corridor",
    cities: [
      "Davao",
      "General Santos",
      "Digos",
      "Tagum",
      "Koronadal",
      "Kidapawan",
    ],
    hub: "Davao",
    characteristics: {
      primary_mode: "bus",
      infrastructure_quality: "excellent",
      tourist_friendly: true,
    },
    recommended_routes: [
      "Davao-General Santos",
      "Davao-Tagum",
      "Davao-Digos",
    ],
  },

  MINDANAO_NORTH: {
    name: "Northern Mindanao Corridor",
    cities: ["Cagayan de Oro", "Iligan", "Butuan", "Valencia", "Malaybalay"],
    hub: "Cagayan de Oro",
    characteristics: {
      primary_mode: "bus",
      infrastructure_quality: "good",
      tourist_friendly: true,
    },
    recommended_routes: [
      "Cagayan de Oro-Iligan",
      "Cagayan de Oro-Valencia",
      "Cagayan de Oro-Butuan",
    ],
  },

  VISAYAS_CENTRAL: {
    name: "Central Visayas",
    cities: [
      "Cebu",
      "Mandaue",
      "Lapu-Lapu",
      "Tagbilaran",
      "Dumaguete",
      "Bohol",
    ],
    hub: "Cebu",
    characteristics: {
      primary_mode: "ferry",
      infrastructure_quality: "excellent",
      tourist_friendly: true,
      island_hopping: true,
      scenic_routes: true,
    },
    recommended_routes: ["Cebu-Bohol", "Cebu-Dumaguete"],
    notes: "Ferry connections are the norm - fast and tourist-friendly",
  },

  VISAYAS_WESTERN: {
    name: "Western Visayas",
    cities: ["Iloilo", "Bacolod", "Kalibo", "Boracay", "Roxas"],
    hub: "Iloilo",
    characteristics: {
      primary_mode: "ferry",
      infrastructure_quality: "good",
      tourist_friendly: true,
      island_hopping: true,
    },
    recommended_routes: ["Iloilo-Bacolod"],
  },

  VISAYAS_EASTERN: {
    name: "Eastern Visayas",
    cities: ["Tacloban", "Ormoc", "Catbalogan"],
    hub: "Tacloban",
    characteristics: {
      primary_mode: "ferry",
      infrastructure_quality: "good",
      tourist_friendly: true,
    },
    recommended_routes: ["Cebu-Ormoc"],
  },

  PALAWAN: {
    name: "Palawan",
    cities: ["Puerto Princesa", "El Nido", "Coron", "San Vicente"],
    hub: "Puerto Princesa",
    characteristics: {
      primary_mode: "van",
      infrastructure_quality: "fair",
      tourist_friendly: true,
      scenic_routes: true,
      limited_service: true,
    },
    notes:
      "Long van rides common. Puerto Princesa to El Nido = 5-6 hours but scenic.",
  },
};

/**
 * Find which regional corridor a city belongs to
 * @param {string} cityName - City to lookup
 * @returns {Object|null} Regional context or null
 */
export const findRegionalContext = (cityName) => {
  if (!cityName) return null;

  const normalizeCity = (city) =>
    city
      .trim()
      .toLowerCase()
      .replace(/\s+/g, " ");
  const normCity = normalizeCity(cityName);

  for (const [regionKey, regionData] of Object.entries(
    REGIONAL_TRANSPORT_CONTEXT
  )) {
    const hasCityMatch = regionData.cities.some(
      (city) => normalizeCity(city) === normCity
    );

    if (hasCityMatch) {
      return {
        regionKey,
        ...regionData,
      };
    }
  }

  return null;
};

/**
 * Check if two cities are in the same regional transport corridor
 * @param {string} city1 - First city
 * @param {string} city2 - Second city
 * @returns {Object} Regional relationship info
 */
export const isRegionalTransportPractical = (city1, city2) => {
  const region1 = findRegionalContext(city1);
  const region2 = findRegionalContext(city2);

  // If either city not in database
  if (!region1 || !region2) {
    return {
      sameRegion: false,
      recommendation:
        "Inter-regional travel - check specific route distances and times",
      bothRegionsFound: !!(region1 && region2),
    };
  }

  // Same region
  if (region1.regionKey === region2.regionKey) {
    // Check if route is in avoid list
    const routeKey = `${city1}-${city2}`;
    const reverseKey = `${city2}-${city1}`;

    const isAvoided = region1.avoid_routes?.some(
      (route) =>
        route.toLowerCase() === routeKey.toLowerCase() ||
        route.toLowerCase() === reverseKey.toLowerCase()
    );

    const isRecommended = region1.recommended_routes?.some(
      (route) =>
        route.toLowerCase() === routeKey.toLowerCase() ||
        route.toLowerCase() === reverseKey.toLowerCase()
    );

    return {
      sameRegion: true,
      region: region1.name,
      isRecommended: isRecommended || false,
      isAvoided: isAvoided || false,
      characteristics: region1.characteristics,
      recommendation: isAvoided
        ? `⚠️ Not recommended within ${region1.name} - consider flying or breaking journey`
        : isRecommended
        ? `✅ ${region1.characteristics.primary_mode} travel recommended within ${region1.name}`
        : `${region1.characteristics.primary_mode} travel available within ${region1.name}`,
      hub: region1.hub,
    };
  }

  // Different regions
  return {
    sameRegion: false,
    region1: region1.name,
    region2: region2.name,
    recommendation: `Inter-regional travel from ${region1.name} to ${region2.name} - check specific route or consider flying`,
    hub1: region1.hub,
    hub2: region2.hub,
  };
};

/**
 * Get transport recommendations for a specific region
 * @param {string} regionKey - Region identifier
 * @returns {Object|null} Region transport recommendations
 */
export const getRegionalRecommendations = (regionKey) => {
  return REGIONAL_TRANSPORT_CONTEXT[regionKey] || null;
};

/**
 * Check if route crosses major geographic boundaries (e.g., inter-island)
 * @param {string} city1 - First city
 * @param {string} city2 - Second city
 * @returns {Object} Boundary crossing info
 */
export const checkGeographicBoundaries = (city1, city2) => {
  const region1 = findRegionalContext(city1);
  const region2 = findRegionalContext(city2);

  if (!region1 || !region2) {
    return { crossesBoundary: false, boundaryType: null };
  }

  const r1Key = region1.regionKey;
  const r2Key = region2.regionKey;

  // Check if crossing major island groups
  const luzonRegions = [
    "METRO_MANILA",
    "LUZON_SOUTH",
    "LUZON_NORTH",
    "LUZON_CENTRAL",
  ];
  const visayasRegions = [
    "VISAYAS_CENTRAL",
    "VISAYAS_WESTERN",
    "VISAYAS_EASTERN",
  ];
  const mindanaoRegions = [
    "MINDANAO_WEST",
    "MINDANAO_SOUTH",
    "MINDANAO_NORTH",
  ];

  const isLuzon1 = luzonRegions.includes(r1Key);
  const isLuzon2 = luzonRegions.includes(r2Key);
  const isVisayas1 = visayasRegions.includes(r1Key);
  const isVisayas2 = visayasRegions.includes(r2Key);
  const isMindanao1 = mindanaoRegions.includes(r1Key);
  const isMindanao2 = mindanaoRegions.includes(r2Key);

  // Check for inter-island crossings
  if ((isLuzon1 && isVisayas2) || (isVisayas1 && isLuzon2)) {
    return {
      crossesBoundary: true,
      boundaryType: "Luzon-Visayas",
      recommendation: "Flight or long ferry required",
    };
  }

  if ((isLuzon1 && isMindanao2) || (isMindanao1 && isLuzon2)) {
    return {
      crossesBoundary: true,
      boundaryType: "Luzon-Mindanao",
      recommendation: "Flight required",
    };
  }

  if ((isVisayas1 && isMindanao2) || (isMindanao1 && isVisayas2)) {
    return {
      crossesBoundary: true,
      boundaryType: "Visayas-Mindanao",
      recommendation: "Flight or ferry required",
    };
  }

  return {
    crossesBoundary: false,
    boundaryType: "same-island-group",
    recommendation: "Ground transport may be available",
  };
};

export default {
  REGIONAL_TRANSPORT_CONTEXT,
  findRegionalContext,
  isRegionalTransportPractical,
  getRegionalRecommendations,
  checkGeographicBoundaries,
};

File: travel-backend/langgraph_agents/data/regional_transport_context.py
"""
Regional Transport Context for the Philippines
Python port of regionalTransportContext.js
Provides intelligence about regional transport corridors and connectivity
"""

from typing import Dict, List, Optional, Any


# Regional transport corridor definitions
REGIONAL_TRANSPORT_CONTEXT = {
    "METRO_MANILA": {
        "name": "Metro Manila & Nearby",
        "cities": ["Manila", "Quezon City", "Makati", "Pasig", "Taguig", "Caloocan"],
        "characteristics": {
            "primary_mode": "within_metro",
            "infrastructure_quality": "excellent",
            "tourist_friendly": True,
            "notes": "Within metro - local transport only",
        },
    },
    
    "LUZON_SOUTH": {
        "name": "Southern Luzon",
        "cities": ["Tagaytay", "Batangas", "Lipa", "Lucena", "Naga", "Legazpi", "Sorsogon", "Daet"],
        "hub": "Manila",
        "characteristics": {
            "primary_mode": "bus",
            "infrastructure_quality": "excellent",
            "tourist_friendly": True,
            "scenic_routes": True,
        },
        "recommended_routes": [
            "Manila-Tagaytay",
            "Manila-Batangas",
            "Manila-Lucena",
            "Manila-Naga",
            "Naga-Legazpi",
        ],
        "avoid_routes": [
            "Manila-Legazpi",  # 10 hours, very long
        ],
    },
    
    "LUZON_NORTH": {
        "name": "Northern Luzon",
        "cities": ["Baguio", "Banaue", "Sagada", "Vigan", "Laoag", "La Union", "Bontoc", "Tuguegarao"],
        "hub": "Manila",
        "characteristics": {
            "primary_mode": "bus",
            "infrastructure_quality": "good",
            "tourist_friendly": True,
            "scenic_routes": True,
            "mountain_routes": True,
        },
        "recommended_routes": [
            "Manila-Baguio", 
            "Baguio-Sagada",
            "Manila-Vigan",
            "Vigan-Laoag",
            "Baguio-Vigan",
        ],
        "avoid_routes": [
            "Manila-Laoag",  # 10 hours, very long
            "Manila-Tuguegarao",  # 10 hours, very long
        ],
    },
    
    "LUZON_CENTRAL": {
        "name": "Central Luzon",
        "cities": ["Angeles", "Clark", "Olongapo", "Subic", "Tarlac", "Cabanatuan", "Baler"],
        "hub": "Manila",
        "characteristics": {
            "primary_mode": "bus",
            "infrastructure_quality": "excellent",
            "tourist_friendly": True,
        },
        "recommended_routes": [
            "Manila-Subic", 
            "Manila-Clark",
            "Manila-Angeles",
            "Manila-Baler",
        ],
    },
    
    "MINDANAO_WEST": {
        "name": "Western Mindanao Corridor",
        "cities": ["Zamboanga", "Pagadian", "Dipolog", "Ozamis"],
        "hub": "Zamboanga",
        "characteristics": {
            "primary_mode": "bus",
            "infrastructure_quality": "good",
            "tourist_friendly": True,
            "scenic_routes": True,
            "coastal_highway": True,
        },
        "recommended_routes": [
            "Zamboanga-Pagadian",
            "Pagadian-Dipolog",
            "Dipolog-Ozamis",
        ],
        "avoid_routes": [
            "Zamboanga-Cagayan de Oro",  # Too long (11+ hours)
        ],
        "notes": "Well-connected coastal corridor. Avoid cross-island routes to Northern Mindanao.",
    },
    
    "MINDANAO_SOUTH": {
        "name": "Southern Mindanao Corridor",
        "cities": ["Davao", "General Santos", "Digos", "Tagum", "Koronadal", "Kidapawan", "Mati", "Cotabato", "Tacurong"],
        "hub": "Davao",
        "characteristics": {
            "primary_mode": "bus",
            "infrastructure_quality": "excellent",
            "tourist_friendly": True,
        },
        "recommended_routes": [
            "Davao-General Santos",
            "Davao-Tagum",
            "Davao-Digos",
            "Davao-Mati",
            "General Santos-Koronadal",
        ],
    },
    
    "MINDANAO_NORTH": {
        "name": "Northern Mindanao Corridor",
        "cities": ["Cagayan de Oro", "Iligan", "Butuan", "Valencia", "Malaybalay", "Camiguin", "Surigao", "Siargao", "Marawi"],
        "hub": "Cagayan de Oro",
        "characteristics": {
            "primary_mode": "bus",
            "infrastructure_quality": "good",
            "tourist_friendly": True,
        },
        "recommended_routes": [
            "Cagayan de Oro-Iligan",
            "Cagayan de Oro-Valencia",
            "Cagayan de Oro-Butuan",
            "Cagayan de Oro-Malaybalay",
            "Butuan-Surigao",
            "Surigao-Siargao",
        ],
    },
    
    "VISAYAS_CENTRAL": {
        "name": "Central Visayas",
        "cities": ["Cebu", "Mandaue", "Lapu-Lapu", "Tagbilaran", "Dumaguete", "Bohol", "Panglao", "Carmen", "Ormoc", "Silay"],
        "hub": "Cebu",
        "characteristics": {
            "primary_mode": "ferry",
            "infrastructure_quality": "excellent",
            "tourist_friendly": True,
            "island_hopping": True,
            "scenic_routes": True,
        },
        "recommended_routes": [
            "Cebu-Bohol", 
            "Cebu-Dumaguete",
            "Cebu-Ormoc",
            "Bacolod-Dumaguete",
            "Tagbilaran-Panglao",
        ],
        "notes": "Ferry connections are the norm - fast and tourist-friendly",
    },
    
    "VISAYAS_WESTERN": {
        "name": "Western Visayas",
        "cities": ["Iloilo", "Bacolod", "Kalibo", "Boracay", "Roxas", "Caticlan"],
        "hub": "Iloilo",
        "characteristics": {
            "primary_mode": "ferry",
            "infrastructure_quality": "good",
            "tourist_friendly": True,
            "island_hopping": True,
        },
        "recommended_routes": [
            "Iloilo-Bacolod",
            "Iloilo-Kalibo",
            "Kalibo-Caticlan",
            "Iloilo-Roxas",
        ],
    },
    
    "VISAYAS_EASTERN": {
        "name": "Eastern Visayas",
        "cities": ["Tacloban", "Ormoc", "Catbalogan", "Maasin"],
        "hub": "Tacloban",
        "characteristics": {
            "primary_mode": "bus",
            "infrastructure_quality": "good",
            "tourist_friendly": True,
        },
        "recommended_routes": [
            "Cebu-Ormoc",
            "Tacloban-Ormoc",
            "Tacloban-Catbalogan",
            "Ormoc-Maasin",
        ],
    },
    
    "PALAWAN": {
        "name": "Palawan",
        "cities": ["Puerto Princesa", "El Nido", "Coron", "San Vicente", "Sabang"],
        "hub": "Puerto Princesa",
        "characteristics": {
            "primary_mode": "van",
            "infrastructure_quality": "fair",
            "tourist_friendly": True,
            "scenic_routes": True,
            "limited_service": True,
        },
        "recommended_routes": [
            "Puerto Princesa-El Nido",
            "Puerto Princesa-Sabang",
            "Puerto Princesa-San Vicente",
        ],
        "notes": "Long van rides common. Puerto Princesa to El Nido = 5-6 hours but scenic.",
    },
    
    "MINDORO": {
        "name": "Mindoro",
        "cities": ["Calapan", "Puerto Galera"],
        "hub": "Calapan",
        "characteristics": {
            "primary_mode": "ferry+bus",
            "infrastructure_quality": "fair",
            "tourist_friendly": True,
            "scenic_routes": True,
        },
        "recommended_routes": [
            "Batangas-Calapan",
            "Batangas-Puerto Galera",
        ],
        "notes": "Access via ferry from Batangas",
    },
}


def find_regional_context(city_name: str) -> Optional[Dict[str, Any]]:
    """
    Find which regional corridor a city belongs to
    
    Args:
        city_name: City to lookup
    
    Returns:
        Regional context dict or None
    """
    if not city_name:
        return None
    
    def normalize_city(city):
        return city.strip().lower().replace("  ", " ")
    
    norm_city = normalize_city(city_name)
    
    for region_key, region_data in REGIONAL_TRANSPORT_CONTEXT.items():
        has_city_match = any(
            normalize_city(city) == norm_city
            for city in region_data["cities"]
        )
        
        if has_city_match:
            return {
                "region_key": region_key,
                **region_data,
            }
    
    return None


def is_regional_transport_practical(city1: str, city2: str) -> Dict[str, Any]:
    """
    Check if two cities are in the same regional transport corridor
    
    Args:
        city1: First city
        city2: Second city
    
    Returns:
        Regional relationship info
    """
    region1 = find_regional_context(city1)
    region2 = find_regional_context(city2)
    
    # If either city not in database
    if not region1 or not region2:
        return {
            "same_region": False,
            "recommendation": "Inter-regional travel - check specific route distances and times",
            "both_regions_found": bool(region1 and region2),
        }
    
    # Same region
    if region1["region_key"] == region2["region_key"]:
        # Check if route is in avoid list
        route_key = f"{city1}-{city2}"
        reverse_key = f"{city2}-{city1}"
        
        avoid_routes = region1.get("avoid_routes", [])
        is_avoided = any(
            route.lower() == route_key.lower() or route.lower() == reverse_key.lower()
            for route in avoid_routes
        )
        
        recommended_routes = region1.get("recommended_routes", [])
        is_recommended = any(
            route.lower() == route_key.lower() or route.lower() == reverse_key.lower()
            for route in recommended_routes
        )
        
        recommendation = (
            f"⚠️ Not recommended within {region1['name']} - consider flying or breaking journey"
            if is_avoided
            else (
                f"✅ {region1['characteristics']['primary_mode']} travel recommended within {region1['name']}"
                if is_recommended
                else f"{region1['characteristics']['primary_mode']} travel available within {region1['name']}"
            )
        )
        
        return {
            "same_region": True,
            "region": region1["name"],
            "is_recommended": is_recommended,
            "is_avoided": is_avoided,
            "characteristics": region1["characteristics"],
            "recommendation": recommendation,
            "hub": region1.get("hub"),
        }
    
    # Different regions
    return {
        "same_region": False,
        "region1": region1["name"],
        "region2": region2["name"],
        "recommendation": f"Inter-regional travel from {region1['name']} to {region2['name']} - check specific route or consider flying",
        "hub1": region1.get("hub"),
        "hub2": region2.get("hub"),
    }


def check_geographic_boundaries(city1: str, city2: str) -> Dict[str, Any]:
    """
    Check if route crosses major geographic boundaries (e.g., inter-island)
    
    Args:
        city1: First city
        city2: Second city
    
    Returns:
        Boundary crossing info
    """
    region1 = find_regional_context(city1)
    region2 = find_regional_context(city2)
    
    if not region1 or not region2:
        return {"crosses_boundary": False, "boundary_type": None}
    
    r1_key = region1["region_key"]
    r2_key = region2["region_key"]
    
    # Define major island groups
    luzon_regions = ["METRO_MANILA", "LUZON_SOUTH", "LUZON_NORTH", "LUZON_CENTRAL"]
    visayas_regions = ["VISAYAS_CENTRAL", "VISAYAS_WESTERN", "VISAYAS_EASTERN"]
    mindanao_regions = ["MINDANAO_WEST", "MINDANAO_SOUTH", "MINDANAO_NORTH"]
    
    is_luzon1 = r1_key in luzon_regions
    is_luzon2 = r2_key in luzon_regions
    is_visayas1 = r1_key in visayas_regions
    is_visayas2 = r2_key in visayas_regions
    is_mindanao1 = r1_key in mindanao_regions
    is_mindanao2 = r2_key in mindanao_regions
    
    # Check for inter-island crossings
    if (is_luzon1 and is_visayas2) or (is_visayas1 and is_luzon2):
        return {
            "crosses_boundary": True,
            "boundary_type": "Luzon-Visayas",
            "recommendation": "Flight or long ferry required",
        }
    
    if (is_luzon1 and is_mindanao2) or (is_mindanao1 and is_luzon2):
        return {
            "crosses_boundary": True,
            "boundary_type": "Luzon-Mindanao",
            "recommendation": "Flight required",
        }
    
    if (is_visayas1 and is_mindanao2) or (is_mindanao1 and is_visayas2):
        return {
            "crosses_boundary": True,
            "boundary_type": "Visayas-Mindanao",
            "recommendation": "Flight or ferry required",
        }
    
    return {
        "crosses_boundary": False,
        "boundary_type": "same-island-group",
        "recommendation": "Ground transport may be available",
    }
